waiter/src/waiter/scheduler.clj

```diff
@@ -638,3 +643,121 @@
+                                                                                     total-timers)})
+                                                            new-instance-ids paired-timers)
+                                                       (concat starting-instance-context-maps)
+                                                       (filterv (fn [{:keys [instance-id timer-contexts]}]
+                                                                  (cond
+                                                                    ;; Stop tracking any instances that disappeared.
+                                                                    (not (contains? known-instance-ids' instance-id)) false
+                                                                    ;; Report startup time for any tracked instances that are now healthy.
+                                                                    (contains? healthy-instance-ids instance-id) (doseq [tc timer-contexts] (timers/stop tc))
+                                                                    ;; Continue tracking all other instances.
+                                                                    :else true))))]
+                 (when (pos? unmatched-instance-count)
+                   (log/warn "Unmatched instances discovered in instance launch metric loop" unmatched-instance-count))
+                 (->ServiceInstancesLaunchTracker service-id known-instance-ids' scheduling-instance-timer-contexts' starting-instance-context-maps'))))))
+
+(defn start-instance-launch-stats-maintainer
+  "go block to collect metrics on the instance launch overhead of waiter services.
+
+  Updated state for the router is read from `router-state-chan`."
+  [router-state-chan service-id->service-description-fn]
+  (let [exit-chan (async/chan 1)
+        query-chan (au/latest-chan)
+        update-state-timer (metrics/waiter-timer "state" "instance-launch-stats-maintainer" "update-state")]
+    (async/go
+      (try
+        (loop [{:keys [known-service-ids service-instance-launch-trackers] :as current-state}
+               {:known-service-ids #{}
+                :service-instance-launch-trackers nil}]
+          (let [new-state
+                (async/alt!
+                  exit-chan
+                  ([message]
+                   (if (= :exit message)
+                     (do
+                       (log/warn "stopping instance-launch-stats-maintainer")
+                       (comment "Return nil to exit the loop"))
+                     current-state))
+
+                  router-state-chan
+                  ([router-state]
+                   (timers/start-stop-time!
+                     update-state-timer
+                     (let [{:keys [service-id->healthy-instances service-id->my-instance->slots service-id->unhealthy-instances]} router-state
+                           incoming-service-ids (set (keys service-id->my-instance->slots))
```

Comment:
Why is `incoming-service-ids` used, as opposed to say `(keys service-id->healthy-instances)`?  Does `incoming-service-ids` contain all services regardless of whether any instances have been allocated at all?

Reply:
The same expression is used in `start-service-chan-maintainer` (state.clj) to detect new services. I think this is what we want so we can detect when a new service has been created.